# 예외처리

### 에러와 예외

- Error
  - 메모리 부족, stack overflow 처럼 발생하면 복구할 수 없는 상황
  - 비 정상적 종료를 막을 수 없기 때문에, 디버깅이 필수적이다.
- Exception
  - 파일이 없거나, 네트워크가 끊기는 등
  - 코드에 의해 수습될 수 있는 비교적 약한 오류들



### 예외 클래스의 계층

|               | Error | Checked exception             | Unchecked exception           |
| ------------- | ----- | ----------------------------- | ----------------------------- |
| 에러의 발생지 |       | 프로그램 외부, 내부 모두 가능 | 프로그램 내부만의 오류이다.   |
| 대처코드      |       | 없으면 컴파일 안됨            | 없어도 컴파일 됨              |
| 예시          |       | IOException, SQLException     | RuntimeError(DivisionZero...) |



### 예외처리하기

- try ~ catch ~

```java
try {
    result = a/b;
}
catch(Exception e) {
    System.out.println("예외발생");
}
```

- try에서 예외가 발생하면 알맞은 catch문으로 가는데, Exception e 자리에 예외의 종류가 들어가고, 발생한 예외가 맞는 catch문으로 가게된다.
- 만약 상위타입의 예외가 먼저 선언되어있다면, 아래 등장하는 catch블록은 실행되지 않는다.
- 가장 최상위 예외인 Exception이 catch문 중 가장 먼저 있다면, 다른 catch블록은 실행되지 않을 것이다.



### finally

- catch 다음에 작성되며, 예외가 발생되지 않거나, 발생되는지 상관없이 finally구문은 실행된다.



### throws

- 함수 뒤에 작성되며, 해당 함수에서 예외가 발생되면, 이 예외를 함수에서 처리하지 않고 함수가 호출된 곳으로 전달하여 예외를 처리하도록 시킨다.

```java
void test() throws Exception1, Exception2 {
    ~~~
}
```

- Unchecked Exception은 오류가 발생해도 컴파일 되기 때문에, throws를 쓰지 않아도 함수가 호출된 곳에서 예외가 처리된다.
